core/capabilities/ccip/ocrimpls/contract_transmitter.go

```diff
@@ -23,54 +23,63 @@ type ToCalldataFunc func(
 	report ocr3types.ReportWithInfo[[]byte],
 	rs, ss [][32]byte,
 	vs [32]byte,
-) (any, error)
+) (string, string, any, error)
+
+func NewToCommitCalldata(defaultMethod, priceOnlyMethod string) ToCalldataFunc {
+	return func(
+		rawReportCtx [2][32]byte,
+		report ocr3types.ReportWithInfo[[]byte],
+		rs, ss [][32]byte,
+		vs [32]byte,
+	) (string, string, any, error) {
+		// Note that the name of the struct field is very important, since the encoder used
+		// by the chainwriter uses mapstructure, which will use the struct field name to map
+		// to the argument name in the function call.
+		// If, for whatever reason, we want to change the field name, make sure to add a `mapstructure:"<arg_name>"` tag
+		// for that field.
+		var info ccipocr3.CommitReportInfo
+		if len(report.Info) != 0 {
+			var err error
+			info, err = ccipocr3.DecodeCommitReportInfo(report.Info)
+			if err != nil {
+				return "", "", nil, err
+			}
+		}
 
-func ToCommitCalldata(
-	rawReportCtx [2][32]byte,
-	report ocr3types.ReportWithInfo[[]byte],
-	rs, ss [][32]byte,
-	vs [32]byte,
-) (any, error) {
-	// Note that the name of the struct field is very important, since the encoder used
-	// by the chainwriter uses mapstructure, which will use the struct field name to map
-	// to the argument name in the function call.
-	// If, for whatever reason, we want to change the field name, make sure to add a `mapstructure:"<arg_name>"` tag
-	// for that field.
-	var info ccipocr3.CommitReportInfo
-	if len(report.Info) != 0 {
-		var err error
-		info, err = ccipocr3.DecodeCommitReportInfo(report.Info)
-		if err != nil {
-			return nil, err
+		method := defaultMethod
+		if len(priceOnlyMethod) > 0 && len(info.MerkleRoots) == 0 && len(info.TokenPrices) > 0 {
```

Comment:
I think we need special handling here since Solana always requires a merkle root but it's zeroed

Reply:
(I have some changes to handle that correctly here https://github.com/smartcontractkit/chainlink/pull/15965/commits/92f87746f305745f7cb339613e9ea59da0ab1b6e https://github.com/smartcontractkit/chainlink/pull/15965/commits/ad19b912297b60c0ab7e9a844a82c99c35c89cca)

Reply:
@amit-momin confirmed that the current implementation works

Reply:
Thanks, but I believe Prashant & Joe have this on their radar.